Use non-persistent connection for MPD

thijsputman opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Instead of relying on the (persistent) connection handling of [`python-mpd2`](https://github.com/Mic92/python-mpd2) (which is not working optimally – see for example https://github.com/Mic92/python-mpd2/issues/31), explicitly connect and disconnect for every command send to MPD.

This resolves https://github.com/home-assistant/core/issues/57633#issuecomment-1483837477 (and probably some related issues) where the MPD integration would get stuck in "limbo": Disconnected _and_ unable to reconnect at the same time. Outside of a full restart of Home Assistant there was no way of recovering from this.
With the three MPD players in my network, this issue would occur once per week; a slight hiccup in the Wi-Fi was often already enough to require a full restart of HA.

This change introduces a bit of overhead, but as the integration is local-polling (ie, nothing actually gets pushed back to us over the persistent connection) I'd say the impact is minimal. The MPD protocol doesn't really appear to be geared towards persistent connections any way.

The best solution would probably be a fix in `python-mpd2` itself, but I'm afraid that's a bit out of scope for me at the moment.

I've tried to keep changes to a minimum (didn't totally succeed 😇) and use an elegant/modern Python approach. I'm not very familiar with Python, so there might be some oddities here and there...

I've been running this code locally for about two months (with three MPD players on the network) and it's been rock solid: Disconnects happen from time to time, the integration always recovers and never gets stuck in "limbo".

In case anyone is wondering: I did initially try to simply brute-force some additional connection attempts into the integration – that didn't work at all: Once `python-mpd2` looses the connection, there doesn't appear to an (obvious) way to get it to reconnect again.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: https://github.com/home-assistant/core/issues/57633#issuecomment-1483837477
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] I have followed the [perfect PR recommendations][perfect-pr]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

thijsputman committed:
Do not use a persistent connection to MPD

In other words, don't rely on the connection management provided by
"python-mpd2". Instead of keeping the connection to MPD open, we
explicitly connect before and disconnect after each command.

There is probably a bit of overhead to this, but as the integration
uses a local-polling approach to begin with, no functionality is lost or
degraded.

This change greatly hardens the MPD integration against both network
issues and problems with the daemon itself. All connection-related
failure modes have effectively been removed.

thijsputman committed:
Update state retrieval methods

Only "async_get_media_image" attempts to connect, all others are either
called from there, or from the main "async_update" method (see previous
commit) which also attempts to connect.

So, this commit mainly revolves around gracefully handling situations
where no connection is available when trying to retrieve MPD state.

Finally, note the removal of "self._commands". This property is only
used at the start of "_async_get_file_image_response" and was thus
changed into a local variable.

thijsputman committed:
Update media-player control methods

These all need to explicitly connect to MPD as part of their flow.

thijsputman committed:
Correct ruff failure (auto-fixed)

emontnemery commented on homeassistant/components/mpd/media_player.py:
```diff
@@ -97,75 +99,94 @@ def __init__(self, server, port, password, name):
+                # MPDClient.connect() doesn't always respect its timeout. To
+                # prevent a deadlock, enforce an additional (slightly longer)
+                # timeout on the coroutine itself.
+                try:
+                    await asyncio.wait_for(
```
We generally use the async_timeout.timeout context manager https://github.com/aio-libs/async-timeout

```py
import async_timeout
...
try:
    async with async_timeout.timeout(self._client.timeout + 5):
```

emontnemery reviewed (changes requested):
I think the change is fine since the implementations is polling, with an interval of 120s.

A few comments though

emontnemery commented on homeassistant/components/mpd/media_player.py:
```diff
@@ -97,75 +99,94 @@ def __init__(self, server, port, password, name):
-        self._status = None
+    @asynccontextmanager
+    async def connection(self):
+        """Handle MPD connect and disconnect."""
```
Maybe add a reference explaining the non-persistent connection is a workaround for https://github.com/Mic92/python-mpd2/issues/31

thijsputman committed:
Use "async_timeout.timeout" context manager

thijsputman committed:
Minor changes

thijsputman commented on homeassistant/components/mpd/media_player.py:
```diff
@@ -97,75 +99,94 @@ def __init__(self, server, port, password, name):
-            # Cleanly disconnect in case connection is not in valid state
-            _LOGGER.debug("Error updating status: %s", error)
-            self._disconnect()
+        return self._is_available not in [None, False]
```
> Why do we need do differentiate between None and False, is it just to control debug logs?

Correct.

Whenever a device initially goes offline (or is offline when HA starts), a single warning is logged. If the device remains offline, all subsequent failed connection attempts raise a debug message instead.

Personal preference of keeping the logs clear of duplicate messages/clutter... I guess the grouping HA does in its log viewer UI more or less achieves the same – if this unwanted / against conventions let me know.

kaitlinsm replied:
Is this conversation resolved? can this fix be merged?

thijsputman replied:
The requested simplification (`return self._is_available is True`) was applied.

The distinction between `False` and `None` (to control debug logging) remains – pending my remark/question:

> Personal preference of keeping the logs clear of duplicate messages/clutter... I guess the grouping HA does in its log viewer UI more or less achieves the same – if this unwanted / against conventions let me know.

kaitlinsm replied:
I guess it's more a question for @emontnemery - personally I think logging a WARN on the first failure and logging a DEBUG on subsequent failures makes sense to me. Noisy logging is unhelpful, and the code looks like it'll behave as designed.

brent20 replied:
Is this resolved so this PR can be merged? Thanks!

emontnemery replied:
Yes, warn on first failure and debug after that is fine, it was just not clear to me that's what was happening.

thijsputman commented:
@emontnemery, are you sure `async_timeout.timeout` is a drop in replacement for `asyncio.wait_for`?

I've been running the updated code for a couple days and I just had a hang of the MPD integration (basically goes unresponsive until I restart the Mopidy deamon on the remote machine).

It seems that it's possible for the remote daemon (Mopidy) to accept an incoming connection and then stall/hang. Appears control is never ceded back to the Python in this case, causing things to hang indefinitely (ignoring all timeouts) on the HA side.

Given my limited Python knowledge: Is it possible that `async_timeout.timeout` is too "kind" in this case (as it doesn't create a new task)?

thijsputman committed:
Merge branch 'dev' into mpd-non-persistent-connection

Rudd-O commented:
Why not fix the mpd integration so that when there is an exception, it reconnects?  This is what I have been doing in my MPRIS integration, with the added safety of a heartbeat plus disconnect and retry when the heartbeat isn't received in the expected time window.

thijsputman commented:
> Why not fix the mpd integration so that when there is an exception, it reconnects? This is what I have been doing in my MPRIS integration, with the added safety of a heartbeat plus disconnect and retry when the heartbeat isn't received in the expected time window.

That's the first thing I tried; never managed to get that to work though: Once the MPD integration and its underlying library start to disagree about the connection state, nothing seems to work any more on my end.

Any command issued from the integration (e.g. "ping", "disconnect") simply results in a "cannot execute this command on a disconnected MPD instance"-response from the library... 

If you've got some example code that works properly on your end, feel free to share! I'd love to give it a try here.

thijsputman committed:
Merge branch 'dev' into mpd-non-persistent-connection

Rudd-O commented:
BTW, you must change uses of `asyncio_timeout` to `asyncio.timeout`.  The code base is currently migrating to that.

thijsputman committed:
Replace "async_timeout" with "asyncio.timeout"

Rudd-O commented on homeassistant/components/mpd/media_player.py:
```diff
@@ -128,7 +127,7 @@ async def connection(self):
                 # timeout on the coroutine itself.
                 try:
-                    async with async_timeout.timeout(self._client.timeout + 5):
+                    async with asyncio.timeout(self._client.timeout + 5):
```
Perfect.

thijsputman replied:
Thanks!

thijsputman committed:
Merge branch 'dev' into mpd-non-persistent-connection

thijsputman committed:
Initialise "self._status" to empty dictionary

Used to be initialised as None, which caused "NoneType is not iterable"
type of issues in case of an unexpected disconnect (at which point
status gets set to None again). Instead of guarding against None
everywhere, using an empty dictionary seemed more prudent...

Furthermore, more cautiously access its members to prevent potential
KeyError-s in similar cases.

thijsputman committed:
Fix livelock in "async_mute_volume()"

This method doesn't need a connection; it calls into two other methods
that actually connect to MPD – attempting to connect from here resulted
in a livelock.

melwinek commented:
When will this be added to HA?

brent20 commented:
Bump- this issue burned me today, was happy to find that a PR exists. Why has this not been merged yet? Thanks!

emontnemery reviewed (approved):
LGTM, thanks @thijsputman 👍 

If you want to change how the log level is decided, please open a follow-up PR.

emontnemery commented:
(Close + Open to rerun CI against the current tip of dev)
